cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java

```diff
@@ -291,58 +279,31 @@ private RuntimeInfo createRuntimeInfo(ProgramId programId, TwillController contr
-          programDescriptor.getProgramId(), controller, queueAdmin, streamAdmin,
-          getFlowletQueues(programDescriptor.getProgramId().getParent(), flowSpec),
-          txExecutorFactory, impersonator
-        );
-        programController = new FlowTwillProgramController(programId, controller, instanceUpdater, runId);
-        break;
-      }
-      case MAPREDUCE:
-        programController = new MapReduceTwillProgramController(programId, controller, runId);
-        break;
-      case WORKFLOW:
-        programController = new WorkflowTwillProgramController(programId, controller, runId);
-        break;
-      case WEBAPP:
-        programController = new WebappTwillProgramController(programId, controller, runId);
-        break;
-      case SERVICE:
-        programController = new ServiceTwillProgramController(programId, controller, runId);
-        break;
-      case WORKER:
-        programController = new WorkerTwillProgramController(programId, controller, runId);
-        break;
+    ProgramRunner programRunner;
+    try {
+      programRunner = programRunnerFactory.create(programId.getType());
+    } catch (IllegalArgumentException e) {
+      // This shouldn't happen. If it happen, it means CDAP was incorrectly install such that some of the program
+      // type is not support (maybe due to version mismatch in upgrade).
+      LOG.error("Unsupported program type {} for program {}. " +
+                  "It is likely caused by incorrect CDAP installation or upgrade to incompatible CDAP version",
+                programId.getType(), programId);
+      return null;
     }
-    return programController == null ? null : programController.startListen();
-  }
 
-  // TODO (terence) : This method is part of the hack mentioned above. It should be removed when
-  // FlowProgramRunner moved to run in AM.
-  private Multimap<String, QueueName> getFlowletQueues(ApplicationId appId, FlowSpecification flowSpec) {
-    // Generate all queues specifications
-    Table<QueueSpecificationGenerator.Node, String, Set<QueueSpecification>> queueSpecs
-      = new SimpleQueueSpecificationGenerator(appId).create(flowSpec);
-
-    // For storing result from flowletId to queue.
-    ImmutableSetMultimap.Builder<String, QueueName> resultBuilder = ImmutableSetMultimap.builder();
-
-    // Loop through each flowlet
-    for (Map.Entry<String, FlowletDefinition> entry : flowSpec.getFlowlets().entrySet()) {
-      String flowletId = entry.getKey();
-      // For each queue that the flowlet is a consumer, store the number of instances for this flowlet
-      for (QueueSpecification queueSpec : Iterables.concat(queueSpecs.column(flowletId).values())) {
-        resultBuilder.put(flowletId, queueSpec.getQueueName());
-      }
+    if (!(programRunner instanceof DistributedProgramControllerFactory)) {
+      // This is also unexpected. If it happen, it means the CDAP core or the runtime provider extension was wrongly
```

Comment:
How can this happen? Guice misconfiguration or something like that? Would it be better to throw an exception so the control operation fails, rather than log an error and continue?